src/mem_fns.rs

```diff
@@ -0,0 +1,659 @@
+) {
+  __aeabi_memset(dest.cast(), byte_count, byte)
+}
+
+/// Copy between non-exclusive regions, prefer [`__aeabi_memset`] if possible.
+///
+/// This function is provided only for API completeness, because in some cases
+/// the compiler might automatically generate a call to this function.
+#[inline]
+#[no_mangle]
+#[instruction_set(arm::a32)]
+#[link_section = ".iwram.__aeabi_memset8"]
+pub unsafe extern "C" fn __aeabi_memset8(
+  dest: *mut u32, byte_count: usize, byte: i32,
+) {
+  __aeabi_memset(dest.cast(), byte_count, byte)
+}
+
+/// Sets all bytes in the region to the `byte` given.
+///
+/// Because of historical reasons, the byte is passed in as an `i32`, but only
+/// the lowest 8 bits are used.
+#[naked]
+#[no_mangle]
+#[instruction_set(arm::a32)]
+#[link_section = ".iwram.__aeabi_memset"]
+pub unsafe extern "C" fn __aeabi_memset(
+  dest: *mut u8, byte_count: usize, byte: i32,
+) {
+  core::arch::asm! {
+    bracer::when!("r1" >=u "#8" [label_id=7] {
+      // duplicate the byte across all of r2 and r3
+      "and    r2, r2, #0xFF",
+      "orr    r2, r2, r2, lsl #8",
+      "orr    r2, r2, r2, lsl #16",
+      "mov    r3, r2",
+
+      // carry/sign test on the address, then do fixup
+      "lsls   r12, r0, #31",
+      "submi  r1, r1, #1",
+      "strbmi r2, [r0], #1",
+      "subcs  r1, r1, #2",
+      "strhcs r2, [r0], #2",
+
+      bracer::when!("r1" >=u "#32" [label_id=8] {
+        bracer::with_pushed_registers!("{{r4-r9}}", {
+          "mov    r4, r2",
+          "mov    r5, r2",
+          "mov    r6, r2",
+          "mov    r7, r2",
+          "mov    r8, r2",
+          "mov    r9, r2",
+          "1:",
+          "subs   r1, r1, #32",
+          "stmge  r0!, {{r2-r9}}",
+          "bgt    1b",
+        }),
+        "bxeq   lr",
+      }),
+
+      // set 4 words
+      "tst    r1, #0b10000",
+      "stmne  r0!, {{r2, r3}}",
+      "stmne  r0!, {{r2, r3}}",
+
+      // set 2 and/or 1 words
+      "lsls   r12, r1, #29",
+      "stmcs  r0!, {{r2, r3}}",
+      "strmi  r2, [r0], #4",
+
+      // set halfword and/or byte
+      "lsls   r12, r1, #31",
+      "strhcs r2, [r0], #2",
+      "strbmi r2, [r0], #1",
+      "bx     lr",
+    }),
+    // byte loop
+    "9:",
+    "subs   r1, r1, #1",
+    "strbcs r2, [r0], #1",
+    "bgt    9b",
+    "bx     lr",
+    options(noreturn)
+  }
+}
+
+/// Copy between non-exclusive regions, prefer [`__aeabi_memset`] if possible.
+///
+/// This is the libc version of a memory set. It's required to return the
+/// `dest` pointer at the end of the call, which makes it need an extra
+/// push/pop compared to a direct call to `__aeabi_memset`.
+///
+/// * **Returns:** The `dest` pointer.
+#[naked]
+#[no_mangle]
+#[instruction_set(arm::a32)]
+#[link_section = ".iwram.memset"]
+pub unsafe extern "C" fn memset(
+  dest: *mut u8, byte: i32, byte_count: usize,
+) -> *mut u8 {
+  core::arch::asm! {
+    bracer::with_pushed_registers!("{{r0, lr}}", {
+      "bl {__aeabi_memset}",
```

Comment:
I'm trying to track a crash that I'm getting after trying to upgrade from v0.10.0 to v0.11.0 and I suspect that this could be caused by this commit not swapping the order of the `byte` and `byte_count` parameters anymore like it did.

mgba trace to demonstrate the issue:

```
# r2 will be set by caller to byte_count=0x3c
02000D48 08033603 00000001 08009ABD 00000000 0000003C 03007BD8 02002636 02000D48 00000002 02001556 00000026 0200292C 03007A08 08009AD5 0803360C cpsr: 2000003F |     002A: lsl r2, r5, #0
# r1 will be set by caller to byte=0
02000D48 08033603 0000003C 08009ABD 00000000 0000003C 03007BD8 02002636 02000D48 00000002 02001556 00000026 0200292C 03007A08 08009AD5 0803360E cpsr: 2000003F |     2100: mov r1, #0
02000D48 00000000 0000003C 08009ABD 00000000 0000003C 03007BD8 02002636 02000D48 00000002 02001556 00000026 0200292C 03007A08 08009AD5 08033610 cpsr: 6000003F | F005FFB3: bl
02000D48 00000000 0000003C 08009ABD 00000000 0000003C 03007BD8 02002636 02000D48 00000002 02001556 00000026 0200292C 03007A08 08038612 08033612 cpsr: 6000003F |     FFB3: bl
02000D48 00000000 0000003C 08009ABD 00000000 0000003C 03007BD8 02002636 02000D48 00000002 02001556 00000026 0200292C 03007A08 08033613 0803957A cpsr: 6000003F |     4778: bx pc
02000D48 00000000 0000003C 08009ABD 00000000 0000003C 03007BD8 02002636 02000D48 00000002 02001556 00000026 0200292C 03007A08 08033613 08039580 cpsr: 6000001F | E51FF004: ldr pc, =0xE51FF004  @ 0x0803957C
02000D48 00000000 0000003C 08009ABD 00000000 0000003C 03007BD8 02002636 02000D48 00000002 02001556 00000026 0200292C 03007A08 08033613 0300076C cpsr: 6000001F | E92D4001: stmdb sp!, {r0,lr}
02000D48 00000000 0000003C 08009ABD 00000000 0000003C 03007BD8 02002636 02000D48 00000002 02001556 00000026 0200292C 03007A00 08033613 03000770 cpsr: 6000001F | EBFFFFD4: bl 0x030006C0
# The length check in __aeabi_memset expects byte_count in r1, but it's in r2: bracer::when!("r1" >=u "#8" [label_id=7] {
02000D48 00000000 0000003C 08009ABD 00000000 0000003C 03007BD8 02002636 02000D48 00000002 02001556 00000026 0200292C 03007A00 03000770 030006C8 cpsr: 6000001F | E3510008: cmp r1, #8
02000D48 00000000 0000003C 08009ABD 00000000 0000003C 03007BD8 02002636 02000D48 00000002 02001556 00000026 0200292C 03007A00 03000770 030006CC cpsr: 8000001F | 3A000021: bcc 0x03000750
# Reverts to the byte loop for 0 bytes
02000D48 00000000 0000003C 08009ABD 00000000 0000003C 03007BD8 02002636 02000D48 00000002 02001556 00000026 0200292C 03007A00 03000770 03000758 cpsr: 8000001F | E2511001: subs r1, r1, #1
02000D48 FFFFFFFF 0000003C 08009ABD 00000000 0000003C 03007BD8 02002636 02000D48 00000002 02001556 00000026 0200292C 03007A00 03000770 0300075C cpsr: 8000001F | 24C02001: strcsb r2, [r0], #1
02000D48 FFFFFFFF 0000003C 08009ABD 00000000 0000003C 03007BD8 02002636 02000D48 00000002 02001556 00000026 0200292C 03007A00 03000770 03000760 cpsr: 8000001F | CAFFFFFC: bgt 0x03000750
02000D48 FFFFFFFF 0000003C 08009ABD 00000000 0000003C 03007BD8 02002636 02000D48 00000002 02001556 00000026 0200292C 03007A00 03000770 03000764 cpsr: 8000001F | E12FFF1E: bx lr
# Returns
02000D48 FFFFFFFF 0000003C 08009ABD 00000000 0000003C 03007BD8 02002636 02000D48 00000002 02001556 00000026 0200292C 03007A00 03000770 03000774 cpsr: 8000001F | E8BD4001: ldmia sp!, {r0,lr}
02000D48 FFFFFFFF 0000003C 08009ABD 00000000 0000003C 03007BD8 02002636 02000D48 00000002 02001556 00000026 0200292C 03007A08 08033613 03000778 cpsr: 8000001F | E12FFF1E: bx lr
02000D48 FFFFFFFF 0000003C 08009ABD 00000000 0000003C 03007BD8 02002636 02000D48 00000002 02001556 00000026 0200292C 03007A08 08033613 08033614 cpsr: 8000003F |     4642: mov r2, r8
02000D48 FFFFFFFF 02000D48 08009ABD 00000000 0000003C 03007BD8 02002636 02000D48 00000002 02001556 00000026 0200292C 03007A08 08033613 08033616 cpsr: 8000003F |     9B0A: ldr r3, [sp, #40]
02000D48 FFFFFFFF 02000D48 0200292C 00000000 0000003C 03007BD8 02002636 02000D48 00000002 02001556 00000026 0200292C 03007A08 08033613 08033618 cpsr: 8000003F |     601A: str r2, [r3, #0]
```

Reply:
Oops, that's definitely bad and would cause crashes if that's the case.

I can have a close look in a few hours.

Reply:
https://github.com/rust-console/gba/commit/05d64354a7949ab154da75ecd6f79120bca6be98 this should solve the issue, does that look right to you @jturcotte ?

Reply:
That fixes it, thanks!